src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -0,0 +1,756 @@
+      }
+      // do not store this next node, we stay at the closest less/equal node to find the equal
+      // node
+    }
+
+    while (currNode != head && comparator.compare(currNode.getValue(), val) == 0) {
+      if (currNode.getValue().equals(pO)) {
+        removeNode(currNode);
+        return true;
+      } else {
+        currNode = currNode.getNext(LEVEL_ONE);
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public int size() {
+    return size;
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return size == 0;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean contains(Object pO) {
+    Preconditions.checkNotNull(pO);
+    @Var Node<T> currNode = head;
+    T val = (T) pO;
+    for (int currLvl = MAX_LEVEL; currLvl >= LEVEL_ONE; currLvl--) {
+      currNode = getClosestLessEqual(currNode, currLvl, val);
+    }
+
+    while (currNode != head
+        && currNode != null
+        && comparator.compare(currNode.getValue(), val) == 0) {
+      if (currNode.getValue().equals(pO)) {
+        return true;
+      } else {
+        currNode = currNode.getNext(LEVEL_ONE);
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public Object[] toArray() {
+    Object[] arr = new Object[size];
+    @Var int i = 0;
+    for (T v : this) {
+      arr[i] = v;
+      i++;
+    }
+    return arr;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T1> T1[] toArray(T1[] a) {
+    Preconditions.checkNotNull(a);
+    List<T1> newList = new ArrayList<>(size());
+
+    for (T v : this) {
+      newList.add((T1) v);
+    }
+
+    return newList.toArray(a);
+  }
+
+  @Override
+  public boolean containsAll(Collection<?> pC) {
+    Preconditions.checkNotNull(pC);
+    for (Object o : pC) {
+      if (!contains(o)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public boolean addAll(Collection<? extends T> pC) {
+    Preconditions.checkNotNull(pC);
+    @Var boolean changed = false;
+    for (T o : pC) {
+      changed = add(o) || changed;
+    }
+    return changed;
+  }
+
+  @Override
+  public boolean removeAll(Collection<?> pC) {
+    Preconditions.checkNotNull(pC);
+    @Var boolean changed = false;
+    for (Object o : pC) {
+      changed = remove(o) || changed;
+    }
+    return changed;
+  }
+
+  @Override
+  public boolean retainAll(Collection<?> pC) {
+    Preconditions.checkNotNull(pC);
+    @Var boolean changed = false;
+    @Var Node<T> currNode = head;
+    while (currNode != null) {
+      @Var Node<T> next = currNode.getNext(LEVEL_ONE);
+      if (next != null && !pC.contains(next.getValue())) {
+        removeNode(next);
+        changed = true;
+      } else {
+        currNode = next;
+      }
+    }
+
+    return changed;
+  }
+
+  @Override
+  public void clear() {
+    head = createHead();
+    size = 0;
+  }
+
+  private Node<T> getNode(int pIndex) {
+    if (pIndex >= size) {
```

Comment:
In cases like this, it also needs to be chcked whether`pIndex` is negative. There exists `Preconditions.checkElementIndex` precisely for this use case.

Reply:
Thanks for the hint!